### ISSUE TITLE
What is the Webpack 2.0 situation in regards to integrating with Aurelia and Angular 2.0?

### ISSUE THREAD

**Issue by dschissler** [webpack/webpack#1262]
Is any thought going into how Webpack will integrate with these next-gen frameworks?

**Comment by sokra** [comment]
Can you give more details?

**Comment by dschissler** [comment]
They are pushing their gulp based jspm cli system (with 'system' something).  It seems that Angular 2.0 might be going the jspm route as well.  It allows HTML and Javascript import syntax to be used and it seems like the most comprehensive ES2015 and ES2016 system around, although they are both pre-release.

Already with Babel in Webpack I'm using the new JS import syntax instead of requires and that is working out well enough.  However, it doesn't seem to support the `import {x, y, z} from 'package'` syntax since its really just mapping it to a CommonJS require.  Then about HTML imports I have no idea about that and maybe it would be covered in a loader.

I'm not well versed in all of the nuances and I'd be waiting for all of the tooling around it to evolve.  Someone at least should be eying Angular 2.0 and Aurelia since they both are both being created with future tech.

I think that Webpack should have a path through all of this new tech so that as much new syntax can be used.

**Comment by sokra** [comment]
Babel supports `import {x, y, z} from 'package'`.

webpack 2 (see webpack-2 branch) supports the ES6 module syntax (`import`, `export`) and the `System` loader API (`System.load`, if the request are statically analysable) natively.

So any library in ES6 syntax is supported (ES6 module syntax enforces static analysablitity).

**Comment by dschissler** [comment]
Ok thanks.  How is Webpack 2 now?  I searched around for information about it and I ended up cowardly retreating.

So then about the ES6 syntax I assumed then that I can do the following:

`import {x, y, z} from 'package'`

and then

```javascript
module.exports = {
  x: 1,
  y: 2
  z: 3
};
```

or with Webpack 2 branch

```javascript
export = {
  x: 1,
  y: 2
  z: 3
};
```

I found this [reddit posting](https://www.reddit.com/r/javascript/comments/2is81v/why_is_not_systemjs_dominating_the_field_of/) helpful but its out of date now with Webpack 2 on the way.

Also what is the ETA on Webpack 2?  Is it working well or are there caveats?

**Comment by dschissler** [comment]
I deleted my last post because it didn't make sense and I was using the new syntax wrong.

Does webpack 2 work with the new string interpolation from `import`?  Also have you used it with Babel and do you somehow disable Babel's interpretation of the `import`, `export`?  There are a few instances where I'm using webpack defined constants for a path (locale and theme root) and Babel doesn't seem to find these and its not resolving the string concatenation.  I'm guessing that webpack 2 would resolve it if they are both constants?  For now I'm just leaving those as good ol' requires.

Then as far as loaders like `bundle` and other splitting mechanisms, does that map to the new ES2015 syntax?  I'm not complaining because I love how I am able to use the bundle loader to wire up the build process while using it for the front end logic of the async locale module.  I'm just trying to move to new syntax now since I'm ready to really start creating a lot of modules in my project.

**Comment by dschissler** [comment]
So I think that however it needs to work that it would be great to be able to able to use the following in newer versions of node that supports the new string templates.

```javascript
import package from `${SOME_ROOT}/package`;
```

So then again the Babel loader `import`, `export` to commonJS translation would be disabled and webpack would handle it.

**Comment by sokra** [comment]
> import package from `${SOME_ROOT}/package`;

I doubt this is valid ES6. `import` only allows constant strings by spec.

**Comment by dschissler** [comment]
Yes but what about the following?

```
require(`${SOME_ROOT}/package`);
```

Is that proper CommonJS?

It seems like its a Webpackism that is very useful and also straight forward.  Might as well offer the ability to resolve string concatenation between constants just like the require.

**Comment by dschissler** [comment]
I suppose that there is always a way around that like putting paths in the lookup paths but its nice to use the ProvidePlugin when its only ever going to used a single time.

**Comment by dschissler** [comment]
Hell, I'd even like to see conditional imports that are not at the top just to keep the same syntax like:


```javascript
if (DEV) {
  import 'debug_widget';
}
```

instead of what I must do now (ok, its not so bad, but...):

```javascript
if (DEV) {
  require('debug_widget');
}
```

Maybe it could be dressed up a bit since it violates the ES2015+ spec but the point its that it keeps with the same new syntax.  Why not `import loader from 'bundle?...'`?

It could be made perhaps somewhat more in line with the ES2015 spec by perhaps making it need to be placed somewhere towards the top before certain kinds of non-import based side effects could get in there.

**Comment by jhnns** [comment]
@dschissler all your example are not statically analyzable, hence no valid ES6 imports. You can only import runtime-depending stuff by using the async `System.load` (afaik).

**Comment by dschissler** [comment]
@jhnns All of my examples are using constants where there is all uppercase with underscores variables.  They are being supplied from the DefinePlugin.  Its nothing more than what webpack already offers and statically figures out, just with the import syntax.  So with `DEV` it looks like `if (TRUE)` or `if (FALSE)`.  Then with the strings it ends up being simple string concatenation and maybe it wouldn't pass NASA but two string literals can very much be statically concatenated.  

Now of course a regex `require` that throws an exception is quite a bit of a different matter.

**Comment by sokra** [comment]
https://people.mozilla.org/~jorendorff/es6-draft.html#sec-imports

```
ImportDeclaration :
  import ImportClause FromClause ;
  import ModuleSpecifier ;

ModuleSpecifier :
  StringLiteral

StringLiteral ::
  " DoubleStringCharacters opt "
  ' SingleStringCharacters opt '

**Comment by use-strict** [comment]
Regarding ES6 imports, what seems to be the heart of the issue is the inability to specify paths that can be rooted to a configurable basePath. Similar to node, you can easily end up with paths like `../../../someModule`, which becomes a hell to manage. As far as I know, ES6 spec doesn't support this, so you can only employ workarounds, some worse than others:

a) Define a module alias for the root path in webpack config and import using module paths:
```
import 'root/someModule'; // with corresponding alias: { root: '/some/base/path'}
```

This gets the job done relatively painless, except your code won't be correctly analyzed by IDEs without some manual help (similar to how we used to do for requirejs)

b) Use another language on top, like TypeScript. If you use `--moduleResolution classic`, you can nicely give paths without a starting `./` or `../` and TypeScript will know to look up from the current path. This also works well with module aliases from a). There is also a path mapping feature, which is not here just yet, but it will be in v1.8 (I think). You will have full IDE support with this approach.

c) Hack away by emulating preprocessing macros. You would need another loader on top or a webpack plugin that removed entire branches. Note that `DefinePlugin` only does dumb replaces, so  you will still end up with something like `if (true)` in your code, causing compilation errors if you used it to wrap import statements. The minify step which is removing dead branches happens too late. This will for sure also make your IDE want to kill you and you might not be able to sleep for a while.

In production, I'm already using TypeScript plus some aliases in webpack config, which works wonderfully.

Regarding conditional imports, this is a tricky subject and most of the time it's a sign you're doing something wrong. There are exceptions, but they are solved in other ways. E.g. platform/browser dependent code, that you'd want removed at compile-time. The feature missing here is preprocessing macros and not a deficiency of the module API. I wouldn't expect to see something like that in the EcmaScript spec any time soon, so you'll probably need to live with some hack like c).

Other times you might find yourself doing battle with the import syntax is when you want something equivalent to require.context() or require-callback syntax. In this cases, this is related to module bundling and not dependency resolving and is solved using either the API provided by webpack (require.context, require.ensure etc) or a separate module loader (ES6 spec provides System.import() promise API)

If you're using require to load locales or configuration or CSS or god knows what, again, ES6 import is probably not the right tool for the job.

**Comment by ivan-kleshnin** [comment]
@use-strict 

> a) Define a module alias for the root path in webpack config and import using module paths:

This does not solve situation for backend web app and standalone scripts (crons).

In my experience the best way to solve this is to add installation step 

bin/install (runned once)
```
# Seed code folders (enables absolute paths like `import x from "frontend"`)
rm -f node_modules/shared   && ln -s ../shared   node_modules/shared
rm -f node_modules/frontend && ln -s ../frontend node_modules/frontend
rm -f node_modules/backend  && ln -s ../backend  node_modules/backend
rm -f node_modules/public   && ln -s ../public   node_modules/public
```

**Comment by use-strict** [comment]
@ivan-kleshnin , the downside to this approach is that symlinks are not cross-platform. In the end, it depends on the project and the team if this is really a problem or not. Though unlikely, you could potentially have naming conflicts with npm packages as well, not just your local folders.

**Comment by ivan-kleshnin** [comment]
@use-strict, right, I forgot about cross-platform aspect, thanks for clarification. 
Doesn't Windows have some symlink alternative yet?!

**Comment by jhnns** [comment]
[Symlinks are supported on Windows](http://www.windows7home.net/how-to-create-symbolic-link-in-windows-7/). I've written [a library](https://github.com/peerigon/link-package) for this use-case, maybe you find it helpful.

**Comment by dschissler** [comment]
What is the fate of my [po-loader](https://www.npmjs.com/package/po-loader)?  It seems fairly popular at usually some thousands of downloads a month.

In particular this function from the documentation does things that I have no idea how to convert to ES6 modules:

```javascript
function getLangLoader(locale) {
  // A runtime exception will be throw every time that the requested locale file 
  // cannot be found. Webpack uses a regular expression to build all locales as 
  // separate bundles. 
  let bundleLoader = require(`bundle!${LOCALE_ROOT}/${locale}/LC_MESSAGES/messages.po`);
  return bundleLoader;
};
```

Additionally I'm completely uncertain how this will fit into Aurelia and Angular 2.

**Comment by jhnns** [comment]
@dschissler `require("bundle!something")` is asynchronous under the hood anyway. That's where `System.import()` comes into play with ES2015 modules. Your solution should still work with `System.import()` afaict.

**Comment by dschissler** [comment]
Ok the lead dev of Aurelia seemed encouraged when I gave him a basic status update on Webpack 2 in regards to it having more ES6 capabilities.  I really think that the kinds of people around here who can write asynchronous concatenated string Javascript in their sleep should be trying to work with that project to ensure that the next generation of big frameworks play well with Webpack 2 before people migrate to System.

**Comment by dschissler** [comment]
There is an effort in Aurelia to create a Webpack loader: https://github.com/aurelia/skeleton-navigation/issues/41#issuecomment-172853531

**Comment by EisenbergEffect** [comment]
The Aurelia team would love to support WebPack (we already support system.js/jspm, require.js/r.js and dojo loader). Our team is less familiar with how WebPack bundles, particularly in the context of ES 2015 modules and the proposed loader API spec. We would love to work with a passionate WebPack/Aurelia community member to help them build out WebPack support and move towards officially supporting this great tool.

**Comment by jhnns** [comment]
@EisenbergEffect what do you need to know exactly? webpack 1 supports CommonJS and AMD and webpack 2 will also understand ES2015 modules + `System.import` for on-demand loading. All module formats are just translated to webpack's internal module format under the hood. So if you generate valid ES2015, there should be no problem.

**Comment by EisenbergEffect** [comment]
@jhnns We've also leveraged loader plugins for loading of text and css, as well as some custom scenarios (basically custom transformed text). Is Webpack flexible enough to handle those types of scenarios? We have a loader abstraction in Aurelia, so we can wrap the System.import api in custom code if needed to handle some of that. I just want to make sure we aren't going to be fighting the system so much that Webpack users find it unnatural.

**Comment by jhnns** [comment]
> Is Webpack flexible enough to handle those types of scenarios

Definitely :grinning:.

So basically, you want to write something like:

```javascript
import { html } from "./my-view.html";
import { css } from "./my-view.css";
```

And with `html` and `css` being the text files as regular strings? In webpack, you would do that with:

```javascript
import { html } from "raw-loader!./my-view.html";
import { css } from "raw-loader!./my-view.css";
```

To avoid repetition (and this rather ugly syntax), most webpack users configure their build so that the `raw-loader` is applied on all HTML- and CSS-files automatically:

```javascript
// webpack.config.js
module.exports = {
    ...
    module: {
        loaders: [
            { test: /\.html$/, loader: "raw-loader" },
            { test: /\.css$/, loader: "raw-loader" },
        ]
    }
};
``` 

However, most webpack users prefer to apply the `html-loader` or the `css-loader` respectively, since these loaders are better suited to handle these file types. They are able to understand the syntax and transform this

```html
<div>
    <h1>Home</h1>
    <img src="assets/img/peerigon.svg">
</div>
```

into that

```
import { img1 } from "./assets/img/peerigon.svg";
export default "<div><h1>Home</h1><img src=\"" + img1 + "\"></div>";
```

which `img1` being the public url (if transformed with the `file-loader`).

Hope that gives you a better understanding about webpack's capabilities. I think it's one of the most flexible build tools around, so I bet your requirements can be met somehow.

**Comment by dschissler** [comment]
Again no single person yet knows exactly how both systems work but perhaps eventually a special HTML loader for Aurelia is desirable.  It could process the HTML into a more efficient template function during build time.  Just from what I know about template systems they eventually make a slim version for the browser that just handles the compiled templates.

I think that ideally Aurelia could be setup with just `import 'aurelia-webpack`.  It seems that using `System` to setup Aurelia as it is currently done would instead be done with Webpack by simply using `<script ...` on the generated Webpack entry point, since that will automatically contain the slim Webpack runtime.

@EisenbergEffect is it possible to have an Aurelia app without any HTML / DOM stuff?

I'm think that it could be best to break it up into first getting the JS initialized and then logging to the console for proof.  Then figuring out if a raw loader will work for the templates or if a Aurelia HTML loader is necessary.

The way that I'd think about Webpack is imagine that you simply attached every class or globally defined module to the `window` object and then referenced it there.  Webpack is essentially just turning a string reference into a integer and storing that in a globally accessible map.

So maybe a `aurelia-loader` isn't the goal.  Usually that is associated with a particular extension, but in this case its just `.js`.

**Comment by dschissler** [comment]
I'm pretty well versed in webpack config but I'm terrible at writing loaders.  If someone could make a typical Aurelia skeleton without HTML stuff that just fires up and logs something to the console then I could try to convert that to a Webpack skeleton.

I'm advocating this approach because I know from experience that getting a template system working can be a huge time sync.  I created the first `nunjucks-loader` for Webpack and it really sucked and took a lot of my peak coffee hours from me and upon request I handed the keys over to a guy who rewrote it and does a good job with it.

*tl;dr* Templates are a huge aggravating time sync and Aurelia + Webpack could move forward more quickly if we ignored the HTML elephant for a moment.
